### ISSUE TITLE
Build failing due to -Werror=deprecated-copy

### ISSUE THREAD

**Issue by GCrispino** [prost-planner/prost#101]
When building the code by running `./build.py` using Manjaro 19.0.1, gcc 9.2.1, the following happens:
```
Building configuration release
-- The C compiler identification is GNU 9.2.1
-- The CXX compiler identification is GNU 9.2.1
-- Check for working C compiler: /usr/bin/gcc
-- Check for working C compiler: /usr/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/g++
-- Check for working CXX compiler: /usr/bin/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found BISON: /usr/bin/bison (found version "3.5.2") 
-- Found FLEX: /usr/bin/flex (found version "2.6.4") 
-- Looking for bdd.h
-- Looking for bdd.h - found
-- Configuring done
-- Generating done
-- Build files have been written to: /home/crispino/Mestrado/Pesquisa/Aplicações/prost/builds/release
make[1]: Entering directory '/home/crispino/Mestrado/Pesquisa/Aplicações/prost/builds/release'
make[2]: Entering directory '/home/crispino/Mestrado/Pesquisa/Aplicações/prost/builds/release'
[  2%] [FLEX][scanner] Building scanner with flex 2.6.4
[  4%] [BISON][parser] Building parser with bison 3.5.2
Scanning dependencies of target rddl-parser
make[2]: Leaving directory '/home/crispino/Mestrado/Pesquisa/Aplicações/prost/builds/release'
make[2]: Entering directory '/home/crispino/Mestrado/Pesquisa/Aplicações/prost/builds/release'
[  6%] Building CXX object rddl_parser/CMakeFiles/rddl-parser.dir/evaluatables.cc.o
[  8%] Building CXX object rddl_parser/CMakeFiles/rddl-parser.dir/instantiator.cc.o
[ 11%] Building CXX object rddl_parser/CMakeFiles/rddl-parser.dir/logical_expressions.cc.o
[ 13%] Building CXX object rddl_parser/CMakeFiles/rddl-parser.dir/preprocessor.cc.o
[ 15%] Building CXX object rddl_parser/CMakeFiles/rddl-parser.dir/probability_distribution.cc.o
[ 17%] Building CXX object rddl_parser/CMakeFiles/rddl-parser.dir/rddl.cc.o
[ 20%] Building CXX object rddl_parser/CMakeFiles/rddl-parser.dir/states.cc.o
[ 22%] Building CXX object rddl_parser/CMakeFiles/rddl-parser.dir/task_analyzer.cc.o
/home/crispino/Mestrado/Pesquisa/Aplicações/prost/src/rddl_parser/task_analyzer.cc: In member function ‘void TaskAnalyzer::analyzeTask(int, int, double)’:
/home/crispino/Mestrado/Pesquisa/Aplicações/prost/src/rddl_parser/task_analyzer.cc:30:43: error: implicitly-declared ‘State& State::operator=(const State&)’ is deprecated [-Werror=deprecated-copy]
   30 |             currentState = State(nextState);
      |                                           ^
In file included from /home/crispino/Mestrado/Pesquisa/Aplicações/prost/src/rddl_parser/task_analyzer.h:4,
                 from /home/crispino/Mestrado/Pesquisa/Aplicações/prost/src/rddl_parser/task_analyzer.cc:1:
/home/crispino/Mestrado/Pesquisa/Aplicações/prost/src/rddl_parser/states.h:22:5: note: because ‘State’ has user-provided ‘State::State(const State&)’
   22 |     State(State const& other) : state(other.state) {}
      |     ^~~~~
/home/crispino/Mestrado/Pesquisa/Aplicações/prost/src/rddl_parser/task_analyzer.cc:32:44: error: implicitly-declared ‘State& State::operator=(const State&)’ is deprecated [-Werror=deprecated-copy]
   32 |             currentState = State(task->CPFs);
      |                                            ^
In file included from /home/crispino/Mestrado/Pesquisa/Aplicações/prost/src/rddl_parser/task_analyzer.h:4,
                 from /home/crispino/Mestrado/Pesquisa/Aplicações/prost/src/rddl_parser/task_analyzer.cc:1:
/home/crispino/Mestrado/Pesquisa/Aplicações/prost/src/rddl_parser/states.h:22:5: note: because ‘State’ has user-provided ‘State::State(const State&)’
   22 |     State(State const& other) : state(other.state) {}
      |     ^~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [rddl_parser/CMakeFiles/rddl-parser.dir/build.make:165: rddl_parser/CMakeFiles/rddl-parser.dir/task_analyzer.cc.o] Error 1
make[2]: Leaving directory '/home/crispino/Mestrado/Pesquisa/Aplicações/prost/builds/release'
make[1]: *** [CMakeFiles/Makefile2:115: rddl_parser/CMakeFiles/rddl-parser.dir/all] Error 2
make[1]: Leaving directory '/home/crispino/Mestrado/Pesquisa/Aplicações/prost/builds/release'
make: *** [Makefile:84: all] Error 2
Building configuration release failed due to CalledProcessError
```

Good news is that I figured out how to fix this. You can either add the `-Wno-error=deprecated-copy` gcc flag to `src/cmake_modules/ProstMacros.cmake`, or fix the code by adding `State& operator=(const State&) = default;` to both `src/rddl_parser/states.h` and `src/search/states.h`.

I'd happily submit a PR for either way, but I'd like to check out with you if any of these approaches looks correct to you.

Cheers!

**Comment by geisserf** [comment]
Thank you for submitting this issue! Fortunately this is reproducible in Ubuntu with gcc-9 as well.

@thomaskeller79 you can install gcc-9 on Ubuntu via:

```
sudo add-apt-repository ppa:ubuntu-toolchain-r/test
sudo apt update
sudo apt install gcc-9
```

The issue is that since c++ 11:

> The generation of the implicitly-defined copy assignment operator is deprecated(since C++11) if T has a user-declared destructor or user-declared copy constructor .(https://en.cppreference.com/w/cpp/language/copy_assignment)

and for states we only have a user-declared copy constructor, but no user-declared copy assignment operator. 

I will submit a pull request on prost-dev soon with a possible fix.

**Comment by GCrispino** [comment]
All right!

**Comment by thomaskeller79** [comment]
Thanks @GCrispino for reporting this issue. 

The issue has been resolved in f8af37b3ddfce7c47bd07380971482a1c66e616e

**Comment by geisserf** [comment]
As a final remark: we decided instead of default declaring copy assignment (`State& operator=(const State&) = default;`) to remove the user declaration of the copy constructor in src/search/states.h and use the default copy constructor instead. The user declaration implemented the default logic anyway, so there was no real reason to have this.
